### ISSUE TITLE
JVM crash related to macOS 10.14+ main thread UI safety

### ISSUE THREAD

**Issue by UberKamek85** [chrislo27/RhythmHeavenRemixEditor#317]
I downloaded the thing, tried to open it via the .jar file ,and it told me to download a JDK, so I did. Then, I tried opening it several times, but it never got past the loading screen. Then it says that "Desktop Launcher" crashed, which I guess is what the computer tried to use to open it.
<img width="494" alt="Screen Shot 2020-05-11 at 7 33 33 PM" src="https://user-images.githubusercontent.com/65200485/81622232-6b362000-93be-11ea-9259-854b3d7494de.png">

**Comment by chrislo27** [comment]
Unfortunately, this is not enough information. Please upload the log file (it is timestamped), found at `~/.rhre3/logs`.

**Comment by CartoonosaurusRex** [comment]
Where do I find it?

**Comment by chrislo27** [comment]
You can find the timestamped log file in your user directory -> `.rhre3` folder -> `logs` folder. You may have to enable [viewing of hidden files/folders in Finder](https://setapp.com/how-to/show-hidden-files-on-mac) first if you cannot find the `.rhre3` folder.

Please note the time that you ran the program as the log files are timestamped with the time of program start. Alternatively, reproduce the error by attempting your same procedure of running the prorgam and upload the most recently generated log file.

**Comment by CartoonosaurusRex** [comment]
I actually tried to open it multiple times, so I'll give you the 3 most recent attempts, as well as the first 2.
[log_2020-05-11_19-17-25.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4612995/log_2020-05-11_19-17-25.txt)
[log_2020-05-11_19-17-39.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4612996/log_2020-05-11_19-17-39.txt)
[log_2020-05-11_19-25-18.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4612997/log_2020-05-11_19-25-18.txt)
[log_2020-05-11_19-28-34.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4612998/log_2020-05-11_19-28-34.txt)
[log_2020-05-11_19-33-23.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4612999/log_2020-05-11_19-33-23.txt)

**Comment by chrislo27** [comment]
Please screenshot the dialog box _Problem Report for DesktopLauncher_ with the Comments opened and the Show Details enabled.

Unfortunately the log file has terminated unexpectedly which may be indicative of a JVM error.

**Comment by CartoonosaurusRex** [comment]
[desktop launcher is drunk.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4613019/desktop.launcher.is.drunk.txt)
<img width="999" alt="Screen Shot 2020-05-11 at 10 20 31 PM" src="https://user-images.githubusercontent.com/65200485/81631385-bb6cac80-93d5-11ea-8620-62a75a9a87db.png">
It would have taken me way too many screenshot to fit everything in Show Details, so I put it in a .txt file instead.

**Comment by chrislo27** [comment]
I think there is a problem with Java 14 given how recent it is (among other things).

I would recommend you to downgrade to Java 8.

First, you will have to uninstall your current Java installation. I am not particularly well-versed with macOS, but I think you can handle the uninstallation.

Second, you should install OpenJDK 8 (LTS) from AdoptOpenJDK. Go to https://adoptopenjdk.net/releases.html, select OpenJDK 8 (LTS) for the version, and HotSpot for the JVM. Then sort by Operating System and select macOS. Download the JRE .pkg (42 MB). In my screenshot it is the third button.

Please let me know how it goes.

![image](https://user-images.githubusercontent.com/6299069/81631925-dd5a3500-93bd-11ea-8149-7d14360842c2.png)

**Comment by CartoonosaurusRex** [comment]
No such luck. Must be something else.

**Comment by CartoonosaurusRex** [comment]
I feel that now is a good time to mention I have only opened it by double-clicking; I tried using the .sh files but Terminal doesn't seem to like them very much.

**Comment by CartoonosaurusRex** [comment]
When I tried, this happened (name censored for anonymity)
![Screen Shot 2020-05-11 at 10 40 37 PM](https://user-images.githubusercontent.com/65200485/81632559-962d6d80-93d8-11ea-94c4-b49b561b79ec.png)

**Comment by chrislo27** [comment]
Hi, I can still tell you are not using Java 8 because the `-version` option failed.

Please restart your computer and try again. To verify, you can open Terminal and enter the following commands:
```
java -version
```
or
```
java --version
```

Please uninstall Java 14 and install OpenJDK 8 before I can help further. Thanks.

**Comment by CartoonosaurusRex** [comment]
I can confirm that I did install OpenJDK8 via the link you provided earlier.
<img width="490" alt="Screen Shot 2020-05-11 at 11 06 39 PM" src="https://user-images.githubusercontent.com/65200485/81634126-4d77b380-93dc-11ea-8647-08f9f517fc21.png">

**Comment by CartoonosaurusRex** [comment]
oh wait, is openjdk8 seperate from java 8
because that would explain a lot

**Comment by chrislo27** [comment]
OpenJDK 8 is just a distribution of Java 8.

I notice you are using `sh` to run the shell file. I would try skipping that by running:
```
$ cd "~/Downloads/Rhythm Heaven Remix Editor/"
$ java -jar -Xmx1024m RHRE.jar
```

For future reference, I have not seen much luck with double-click-to-run on macOS, nor is it an officially recommended method of launching the program.

**Comment by CartoonosaurusRex** [comment]
<img width="542" alt="Screen Shot 2020-05-11 at 11 18 53 PM" src="https://user-images.githubusercontent.com/65200485/81634881-d93e0f80-93dd-11ea-8f7f-ad528e96d6a0.png">
think i should be more specific with the directory?

**Comment by chrislo27** [comment]
Please run my commands exactly to expedite this issue. You forgot a forward slash / between the tilde ~ and Downloads.

**Comment by CartoonosaurusRex** [comment]
nope, still nothing.

**Comment by chrislo27** [comment]
Please run:
```
$ cd "/Users/alexash/Downloads/Rhythm Heaven Remix Editor/"
$ pwd
```
and check that the output of `pwd` is the same path you `cd`'d into. Then run the existing `java` command from above.

**Comment by CartoonosaurusRex** [comment]
I tried it, and it it opened, but then it crashed on the loading screen _again_. I'm getting very annoyed at this, and I imagine you are as well.

**Comment by chrislo27** [comment]
Now that it is running with Java 8, please provide the RHRE log file again and the details text of the _Problem Report for DesktopLauncher_,

Unfortunately, this is looking increasingly like a problem with your computer and not with RHRE, since RHRE starts and no exceptions are thrown that I have caused. 

Thanks for your patience.

**Comment by CartoonosaurusRex** [comment]
No problem man. If it's my computer, I guess I'm gonna have to find another way to make remixes.
[desktop is drunk again.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4613247/desktop.is.drunk.again.txt)
[log_2020-05-11_23-30-07.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4613238/log_2020-05-11_23-30-07.txt)
[log_2020-05-11_23-34-36.txt](https://github.com/chrislo27/RhythmHeavenRemixEditor/files/4613239/log_2020-05-11_23-34-36.txt)

**Comment by chrislo27** [comment]
Some questions:  
- Do you have focus on the RHRE window when it launches?
- Do you ever switch focus from the RHRE window when it launches? (i.e., do you switch windows while it loads)

**Comment by CartoonosaurusRex** [comment]
Yes, I focus on it, and no, I don't switch focus. Guess my computer's just not fit for the program .

**Comment by chrislo27** [comment]
After investigating this problem further, it seems a change happened in macOS 10.14 related to the UI and what is allowed to update the window on what thread. If you search "NSWindow drag regions should only be invalidated on the Main Thread!" you'll find a lot of weird errors from _many_ varying sources. This may not be an easily resolvable problem, especially since RHRE launches and even forms a window when it starts.

This issue will likely be open for a while so please hold tight. I am asking colleagues if they can help test some things for me.  Again, thanks so much for your patience and cooperation.

**Comment by CartoonosaurusRex** [comment]
Again, no problem man.

**Comment by chrislo27** [comment]
I've been working at this issue for a little while and would greatly appreciate your help if you could test this.

Please download and run [Bouncy Road Mania](https://github.com/chrislo27/BouncyRoadMania/releases), which is a game I wrote heavily based on RHRE but actually uses a different backend (GLFW). If this backend works for you, then I think I will migrate RHRE to this new backend.

**Comment by UberKamek85** [comment]
Bouncy Road Mania works fine on my end!

**Comment by chrislo27** [comment]
That's great news! I will start to migrate RHRE to use LWJGL3 then. This may take a few days since it is a planned major update (v3.20.0).

Thanks for your help. I'll keep you posted.

**Comment by chrislo27** [comment]
Hi, I would really appreciate it if you could help me test the LWJGL3 backend.

There's a snapshot of the program available here [v3.20.0-SNAPSHOT_20200512a](https://github.com/chrislo27/RHRE-snapshots/releases/tag/v3.20.0-SNAPSHOT_20200512a). It is the same as v3.19.5 but with the new LWJGL3 backend.

Thanks!

**Comment by UberKamek85** [comment]
Just launched it, and it's so far, so good!

**Comment by UberKamek85** [comment]
actually making a pretty good remix rn

**Comment by UberKamek85** [comment]
Yeah, works great on my end!

**Comment by chrislo27** [comment]
This issue has been resolved as of [v3.20.0](https://github.com/chrislo27/RhythmHeavenRemixEditor/releases/tag/v3.20.0). Thanks @UberKamek85 for your help!
